src/pygenstability/data_clustering.py

```diff
@@ -0,0 +1,286 @@
+        log_scale=True,
+        scales=None,
+        n_tries=100,
+        with_NVI=True,
+        n_NVI=20,
+        with_postprocessing=True,
+        with_ttprime=True,
+        with_spectral_gap=True,
+        exp_comp_mode="spectral",
+        result_file="results.pkl",
+        n_workers=4,
+        tqdm_disable=False,
+        with_optimal_scales=True,
+        optimal_scales_kwargs=None,
+        ms_method="louvain",
+        constructor_kwargs=None,
+    ):
+
+        # initialise parameters for graph construction
+        super().__init__(
+            metric=metric,
+            method=graph_method,
+            k=k,
+            delta=delta,
+            distance_threshold=distance_threshold,
+        )
+
+        # initialise parameters for PyGenStability
+        self.constructor = constructor
+        self.min_scale = min_scale
+        self.max_scale = max_scale
+        self.n_scale = n_scale
+        self.log_scale = log_scale
+        self.scales = scales
+        self.n_tries = n_tries
+        self.with_NVI = with_NVI
+        self.n_NVI = n_NVI
+        self.with_postprocessing = with_postprocessing
+        self.with_ttprime = with_ttprime
+        self.with_spectral_gap = with_spectral_gap
+        self.exp_comp_mode = exp_comp_mode
+        self.result_file = result_file
+        self.n_workers = n_workers
+        self.tqdm_disable = tqdm_disable
+        self.with_optimal_scales = with_optimal_scales
+        self.optimal_scales_kwargs = optimal_scales_kwargs
+        self.MS_method = ms_method
+        self.constructor_kwargs = constructor_kwargs
+
+        # attributes
+        self.adjacency_ = csr_matrix
+        self.results_ = {}
+        self.labels_ = []
+
+    def fit(self, X):
+
+        # construct graph
+        self.adjacency_ = csr_matrix(super().fit(X))
+
+        # adapt optimal scales parameters
+        if self.optimal_scales_kwargs is None:
+            self.optimal_scales_kwargs = {"kernel_size": int(0.2 * self.n_scale)}
+
+        # run PyGenStability
+        self.results_ = pgs_run(
+            self.adjacency_,
+            self.constructor,
+            self.min_scale,
+            self.max_scale,
+            self.n_scale,
+            self.log_scale,
+            self.scales,
+            self.n_tries,
+            self.with_NVI,
+            self.n_NVI,
+            self.with_postprocessing,
+            self.with_ttprime,
+            self.with_spectral_gap,
+            self.exp_comp_mode,
+            self.result_file,
+            self.n_workers,
+            self.tqdm_disable,
+            self.with_optimal_scales,
+            self.optimal_scales_kwargs,
+            self.MS_method,
+            self.constructor_kwargs,
+        )
+
+        # store labels of robust partitions
+        self._postprocess_selected_partitions()
+
+        return self.results_
+
+    def _postprocess_selected_partitions(self):
+        """Postprocess selected partitions."""
+
+        self.labels_ = []
```

Comment:
I'm not sure why you need to store the labels, could you not just return labels from this function, which you return then in line 206? You erase it anyway in this function

Reply:
you are right, I will define the labels as a property of the class.